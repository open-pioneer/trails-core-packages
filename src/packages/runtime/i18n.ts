// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createIntl, createIntlCache, IntlFormatters, IntlShape } from "@formatjs/intl";
import { ErrorId } from "./errors";
import { createLogger, Error } from "@open-pioneer/core";
import { ApplicationMetadata } from "./metadata";
const LOG = createLogger("runtime:i18n");

/**
 * Represents i18n info for the entire application.
 * Currently not exposed to user code.
 */
export interface AppI18n {
    /** Chosen locale */
    readonly locale: string;

    /** Supported locales from app metadata. */
    readonly supportedMessageLocales: string[];

    /** True if the locale can be used in this application (i.e. if there are any messages). */
    supportsLocale(locale: string): boolean;

    /** Given the package name, constructs a package i18n instance. */
    createPackageI18n(packageName: string): PackageIntl;
}

/**
 * Gives access to the package's i18n messages for the current locale.
 *
 * See also https://formatjs.io/docs/intl
 */
export type PackageIntl = Pick<IntlShape, "locale" | "timeZone"> & IntlFormatters<string>;

export function createPackageIntl(locale: string, messages: Record<string, string>) {
    const cache = createIntlCache();
    return createIntl(
        {
            locale,
            messages
        },
        cache
    );
}

/**
 * Initializes the application's locale and fetches the appropriate i18n messages.
 */
export async function initI18n(
    appMetadata: ApplicationMetadata | undefined,
    forcedLocale: string | undefined
): Promise<AppI18n> {
    const messageLocales = appMetadata?.locales ?? [];
    const userLocales = getBrowserLocales();
    if (LOG.isDebug()) {
        LOG.debug(
            `Attempting to pick locale for user (locales: ${userLocales.join(
                ", "
            )}) from app (supported locales: ${messageLocales.join(
                ", "
            )}) [forcedLocale=${forcedLocale}].`
        );
    }

    const i18nConfig = new I18nConfig(messageLocales);
    const { locale, messageLocale } = i18nConfig.pickSupportedLocale(forcedLocale, userLocales);

    if (LOG.isDebug()) {
        LOG.debug(`Using locale '${locale}' with messages from locale '${messageLocale}'.`);
    }

    let messages: Record<string, Record<string, string>>;
    if (messageLocales.includes(messageLocale)) {
        try {
            messages = (await appMetadata?.loadMessages?.(messageLocale)) ?? {};
        } catch (e) {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to load messages for locale '${messageLocale}'.`,
                {
                    cause: e
                }
            );
        }
    }
    return {
        locale,
        supportedMessageLocales: messageLocales,
        supportsLocale(locale) {
            return i18nConfig.supportsLocale(locale);
        },
        createPackageI18n(packageName) {
            const packageMessage = messages?.[packageName] ?? {};
            return createPackageIntl(locale, packageMessage);
        }
    };
}

/** Creates an empty i18n instance, e.g. for tests. */
export function createEmptyI18n(locale = "en"): PackageIntl {
    return createPackageIntl(locale, {});
}

export interface LocalePickResult {
    /**
     * The actual locale (e.g. en-US) for number and date formatting etc.
     */
    locale: string;

    /**
     * The locale identifier to load messages for. E.g. "en".
     */
    messageLocale: string;
}

/**
 * Picks a locale for the app. Exported for tests.
 */
export class I18nConfig {
    private appLocales: string[];

    /**
     * @param appLocales Locales the app has i18n messages for (e.g. "en", "de")
     */
    constructor(appLocales: string[]) {
        this.appLocales = appLocales;
    }

    /**
     * @param forcedLocale Optional forced locale (must be satisfied)
     * @param userLocales Locales requested by the user's browser
     */
    pickSupportedLocale(forcedLocale: string | undefined, userLocales: string[]): LocalePickResult {
        const { appLocales } = this;
        // Attempt to satisfy forced locale
        if (forcedLocale) {
            const result = this.pickImpl([forcedLocale]);
            if (!result) {
                const localesList = appLocales.join(", ");
                throw new Error(
                    ErrorId.UNSUPPORTED_LOCALE,
                    `Locale '${forcedLocale}' cannot be forced because it is not supported by the application.` +
                        ` Supported locales are ${localesList}.`
                );
            }
            if (result.locale !== result.messageLocale) {
                LOG.warn(
                    `Non-exact match for forced locale '${forcedLocale}': using messages from '${result.messageLocale}'.`
                );
            }
            return result;
        }

        // Match preferred locale
        const supportedLocale = this.pickImpl(userLocales);
        if (supportedLocale) {
            return supportedLocale;
        }

        // Fallback: Most preferred locale (for dates etc.), but some of our messages
        return {
            locale: userLocales[0] ?? "en",
            messageLocale: appLocales[0] ?? "en"
        };
    }

    supportsLocale(locale: string): boolean {
        return this.pickImpl([locale]) != null;
    }

    private pickImpl(requestedLocales: string[]) {
        const appLocales = this.appLocales;
        for (const requestedLocale of requestedLocales) {
            // try exact match
            if (appLocales.includes(requestedLocale)) {
                return { messageLocale: requestedLocale, locale: requestedLocale };
            }

            // try plain language tag (e.g. "en") as fallback
            const plainLanguage = requestedLocale.match(/^([a-z]+)/i)?.[1];
            if (plainLanguage && appLocales.includes(plainLanguage)) {
                return { messageLocale: plainLanguage, locale: requestedLocale };
            }
        }
        return undefined;
    }
}

/**
 * Returns locales supported by the browser, in order of preference (preferred first).
 *
 * See also https://developer.mozilla.org/en-US/docs/Web/API/Navigator/languages
 */
export function getBrowserLocales(): string[] {
    if (window.navigator.languages && window.navigator.languages.length) {
        return Array.from(window.navigator.languages);
    }
    return [window.navigator.language];
}
