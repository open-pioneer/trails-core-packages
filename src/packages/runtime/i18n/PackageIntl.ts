// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import {
    createIntl,
    createIntlCache,
    IntlConfig,
    IntlErrorCode,
    IntlFormatters,
    IntlShape,
    MessageDescriptor,
    OnErrorFn
} from "@formatjs/intl";
import {
    type FormatXMLElementFn,
    type Options as IntlMessageFormatOptions,
    type PrimitiveType
} from "intl-messageformat"; // not a dependency, this is a dependency of formatjs
import { createElement, Fragment, ReactNode, Children } from "react";

type BaseIntl = Pick<IntlShape, "locale" | "timeZone"> & IntlFormatters<string>;

/**
 * Gives access to the package's i18n messages for the current locale.
 *
 * See also https://formatjs.io/docs/intl
 */
export type PackageIntl = BaseIntl & PackageIntlExtensions;

/**
 * Trails specific extensions to formatjs' intl API.
 */
export interface PackageIntlExtensions {
    /**
     * Similar to `formatMessage()`, but supports _rich text formatting_.
     *
     * React nodes are supported as input values, and custom tags can be implemented as react components.
     * Note that the output is always a single react node.
     *
     * *Example with react node as value:*
     *
     * ```tsx
     * function Example() {
     *   // Given i18n message "Hello, {name}!"
     *   // replaces 'name' with the given react node:
     *   const message = intl.formatRichMessage({ id: "foo"}, {
     *     name: <FancyUserName />
     *   });
     *   return <Box>{message}</Box>;
     * }
     * ```
     *
     * *Example with basic formatting:*
     *
     * ```tsx
     * function Example() {
     *   // Given i18n message "Hello, <strong>{name}</strong>!"
     *   // renders with actual <strong> html node:
     *   return <Box>{intl.formatRichMessage({ id: "foo"}, { name: "User" })}</Box>;
     * }
     * ```
     *
     * Note that only a few basic formatting tags are predefined ("b", "strong", "i", "em", "code", "br").
     * If you need more advanced tags, you can define your own, see below.
     *
     * *Example with custom tag:*
     *
     * ```tsx
     * function Example() {
     *   // Given i18n message "Open <foo>the door</foo>!",
     *   // renders 'foo' using the formatter function below:
     *   const message = intl.formatRichMessage({ id: "foo"}, {
     *     foo: (parts) => <FancyTag>{parts}</FancyTag>
     *   });
     *   return <Box>{message}</Box>;
     * }
     * ```
     */
    formatRichMessage: (
        descriptor: MessageDescriptor,
        values?: Record<string, RichTextValue>,
        opts?: IntlMessageFormatOptions
    ) => ReactNode;
}

/**
 * Value types supported when rendering rich text messages.
 * Primitive values (or single react nodes) are used for normal values.
 * Functions are used to define how tags should be rendered.
 *
 * See {@link PackageIntlExtensions.formatRichMessage}
 */
export type RichTextValue = PrimitiveType | ReactNode | FormatXMLElementFn<ReactNode>;

/**
 * Creates a package intl instance with the given locale and messages.
 *
 * @internal
 */
export function createPackageIntl(
    locale: string,
    messages: Record<string, string>,
    options?: {
        /** Used during testing to suppress console messages. */
        suppressNotFoundError?: boolean;

        /** Also used during testing */
        defaultLocale?: string;
    }
): PackageIntl {
    const cache = createIntlCache();
    const config: IntlConfig = {
        locale,
        messages
    };
    if (options?.suppressNotFoundError) {
        config.onError = ignoreMissingTranslationError;
    }
    if (options?.defaultLocale) {
        config.defaultLocale = options.defaultLocale;
    }

    const intl = createIntl(config, cache) as unknown as PackageIntl;
    intl.formatRichMessage = formatRichMessage.bind(undefined, intl);
    return intl;
}

/** Creates an empty i18n instance, e.g. for tests. */
export function createEmptyPackageIntl(locale = "en"): PackageIntl {
    return createPackageIntl(locale, {});
}

const RICH_TEXT_TAGS = Object.freeze({
    "b": renderTag("b"),
    "strong": renderTag("strong"),
    "i": renderTag("i"),
    "em": renderTag("em"),
    "code": renderTag("code"),
    "br": renderTag("br")
});

function formatRichMessage(
    intl: PackageIntl,
    descriptor: MessageDescriptor,
    values: Record<string, RichTextValue> | undefined,
    opts: IntlMessageFormatOptions | undefined
): ReactNode {
    values = {
        ...RICH_TEXT_TAGS,
        ...values
    };

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const formatResult: any = intl.formatMessage(descriptor, preprocessValues(values) as any, opts);
    return createElement(Fragment, undefined, ...formatResult);
}

function preprocessValues(values: Record<string, RichTextValue> | undefined) {
    if (!values) {
        return undefined;
    }

    const fixedValues: Record<string, RichTextValue> = {};
    for (const [key, valueOrFn] of Object.entries(values)) {
        let fixedValue = valueOrFn;
        if (typeof valueOrFn === "function") {
            fixedValue = (parts) => toKeyedChildren(valueOrFn(parts));
        }
        fixedValues[key] = fixedValue;
    }
    return fixedValues;
}

function renderTag(tag: string): FormatXMLElementFn<ReactNode> {
    return function renderRichTag(parts) {
        return createElement(tag, undefined, ...parts);
    };
}

// Seems to be necessary to avoid "Each child in a list should have a unique "key" prop." error
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function toKeyedChildren(children: any) {
    return Children.toArray(children);
}

/** Hides missing translation errors during tests */
const ignoreMissingTranslationError: OnErrorFn = (err) => {
    if (err.code === IntlErrorCode.MISSING_TRANSLATION) {
        return;
    }

    console.error(err);
};
