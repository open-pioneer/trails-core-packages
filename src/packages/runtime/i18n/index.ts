// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export { type AppIntl, initI18n } from "./AppIntl";
export { type PackageIntl, createPackageIntl, createEmptyPackageIntl } from "./PackageIntl";
export { type LocalePickResult, I18nConfig, getBrowserLocales } from "./pick";
