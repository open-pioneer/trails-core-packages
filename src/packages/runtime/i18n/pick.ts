// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createLogger, Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
const LOG = createLogger("runtime:i18n");

export interface LocalePickResult {
    /**
     * The actual locale (e.g. en-US) for number and date formatting etc.
     */
    locale: string;

    /**
     * The locale identifier to load messages for. E.g. "en".
     */
    messageLocale: string;
}

/**
 * Picks a locale for the app.
 */
export class I18nConfig {
    private appLocales: string[];

    /**
     * @param appLocales Locales the app has i18n messages for (e.g. "en", "de")
     */
    constructor(appLocales: string[]) {
        this.appLocales = appLocales;
    }

    /**
     * @param forcedLocale Optional forced locale (must be satisfied)
     * @param userLocales Locales requested by the user's browser
     */
    pickSupportedLocale(forcedLocale: string | undefined, userLocales: string[]): LocalePickResult {
        const { appLocales } = this;
        // Attempt to satisfy forced locale
        if (forcedLocale) {
            const result = this.pickImpl([forcedLocale]);
            if (!result) {
                const localesList = appLocales.join(", ");
                throw new Error(
                    ErrorId.UNSUPPORTED_LOCALE,
                    `Locale '${forcedLocale}' cannot be forced because it is not supported by the application.` +
                        ` Supported locales are ${localesList}.`
                );
            }
            if (result.locale !== result.messageLocale) {
                LOG.warn(
                    `Non-exact match for forced locale '${forcedLocale}': using messages from '${result.messageLocale}'.`
                );
            }
            return result;
        }

        // Match preferred locale
        const supportedLocale = this.pickImpl(userLocales);
        if (supportedLocale) {
            return supportedLocale;
        }

        // Fallback: Most preferred locale (for dates etc.), but some of our messages
        return {
            locale: userLocales[0] ?? "en",
            messageLocale: appLocales[0] ?? "en"
        };
    }

    supportsLocale(locale: string): boolean {
        return this.pickImpl([locale]) != null;
    }

    private pickImpl(requestedLocales: string[]) {
        const appLocales = this.appLocales;
        for (const requestedLocale of requestedLocales) {
            // try exact match
            if (appLocales.includes(requestedLocale)) {
                return { messageLocale: requestedLocale, locale: requestedLocale };
            }

            // try plain language tag (e.g. "en") as fallback
            const plainLanguage = requestedLocale.match(/^([a-z]+)/i)?.[1];
            if (plainLanguage && appLocales.includes(plainLanguage)) {
                return { messageLocale: plainLanguage, locale: requestedLocale };
            }
        }
        return undefined;
    }
}

/**
 * Returns locales supported by the browser, in order of preference (preferred first).
 *
 * See also https://developer.mozilla.org/en-US/docs/Web/API/Navigator/languages
 */
export function getBrowserLocales(): string[] {
    if (window.navigator.languages && window.navigator.languages.length) {
        return Array.from(window.navigator.languages);
    }
    return [window.navigator.language];
}
