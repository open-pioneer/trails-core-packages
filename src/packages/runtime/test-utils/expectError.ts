// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export function expectError(impl: () => unknown): Error {
    try {
        impl();
        throw new Error("expected error!");
    } catch (e) {
        if (e instanceof Error) {
            return e;
        }
        throw new Error("unexpected error value, not an instance of Error");
    }
}

export function expectAsyncError(impl: () => Promise<unknown>): Promise<Error> {
    const promise = impl();
    return promise.then(
        () => {
            throw new Error("expected error!");
        },
        (e) => {
            if (e instanceof Error) {
                return e;
            }
            throw new Error("unexpected error value, not an instance of Error");
        }
    );
}
