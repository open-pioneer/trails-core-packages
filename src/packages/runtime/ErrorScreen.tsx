// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0

import {
    Box,
    VStack,
    Heading,
    AlertIcon,
    AlertTitle,
    AlertDescription,
    Alert,
    Text
} from "@open-pioneer/chakra-integration";
import { PackageIntl } from "./i18n";

const MESSAGES_DE = {
    "title": "Anwendungsstart fehlgeschlagen",
    "alertTitle": "Fehler",
    "alertDescription": "Leider ist beim Start der Anwendung ein Fehler aufgetreten.",

    "details.title": "Fehlerdetails",
    "details.inspect": "Konsole Ã¶ffnen, um weitere Details zu sehen.",
    "details.errorMessage": "Fehler: {message}"
};

const MESSAGES_EN = {
    "title": "Application start failed",
    "alertTitle": "Error",
    "alertDescription": "Unfortunately an error occurred during application start.",

    "details.title": "Error details",
    "details.inspect": "Inspect the console to see more details.",
    "details.errorMessage": "Error: {message}"
};

export const MESSAGES_BY_LOCALE: Record<"en" | "de", Record<string, string>> = {
    "en": MESSAGES_EN,
    "de": MESSAGES_DE
};

const isDev = import.meta.env.DEV;

// eslint-disable-next-line @typescript-eslint/no-explicit-any

export function ErrorScreen(props: { intl: PackageIntl; error: Error }) {
    const intl = props.intl;
    return (
        <Box>
            <VStack padding={4} width="100%" height="100%">
                <Heading size="md">{intl.formatMessage({ id: "title" })}</Heading>
                <Alert status="error" maxWidth={550} borderRadius="4px" mt={2}>
                    <AlertIcon />
                    <AlertTitle>{intl.formatMessage({ id: "alertTitle" })}</AlertTitle>
                    <AlertDescription>
                        {intl.formatMessage({ id: "alertDescription" })}
                    </AlertDescription>
                </Alert>

                {isDev && (
                    <Box
                        className="error-details"
                        mt={4}
                        p={2}
                        maxWidth="100%"
                        border={"dotted 2px"}
                        borderColor={"red.600"}
                        borderRadius={"4px"}
                    >
                        <Text color={"red.600"} fontWeight={"bold"}>
                            {intl.formatMessage({ id: "details.title" }) + ":"}
                        </Text>
                        <ErrorRenderer {...props} />
                    </Box>
                )}
            </VStack>
        </Box>
    );
}

/**
 * Renders a single error instance.
 */
function ErrorRenderer(props: { error: Error; intl: PackageIntl }) {
    const { error, intl } = props;
    const message = error?.message || "Unknown";
    const stackTrace = error?.stack;
    const stackTraceDisplay = stackTrace && (
        <>
            <Box
                p="2"
                mt={1}
                mb={3}
                maxHeight="500px"
                whiteSpace="pre"
                wordBreak="keep-all"
                overflow="auto"
                border="1px solid grey"
                borderRadius="4px"
                color="gray.700"
                fontSize="12px"
                fontFamily="monospace"
            >
                {stackTrace}
            </Box>
            {intl.formatMessage({ id: "details.inspect" })}
        </>
    );
    return (
        <Box p={2} className={"error-screen-stack-trace"}>
            <Text fontStyle={"italic"}>
                {intl.formatMessage({ id: "details.errorMessage" }, { message })}
            </Text>
            {stackTraceDisplay}
        </Box>
    );
}
