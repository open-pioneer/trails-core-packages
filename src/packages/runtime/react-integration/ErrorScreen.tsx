// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Alert, Box, Heading, Text, VStack } from "@chakra-ui/react";
import { PackageIntl } from "../i18n";

const MESSAGES_DE = {
    "title": "Anwendungsstart fehlgeschlagen",
    "alertTitle": "Fehler",
    "alertDescription": "Leider ist beim Start der Anwendung ein Fehler aufgetreten.",

    "details.title": "Fehlerdetails",
    "details.inspect": "Konsole Ã¶ffnen, um weitere Details zu sehen.",
    "details.errorMessage": "Fehler: {message}"
};

const MESSAGES_EN = {
    "title": "Application start failed",
    "alertTitle": "Error",
    "alertDescription": "Unfortunately an error occurred during application start.",

    "details.title": "Error details",
    "details.inspect": "Inspect the console to see more details.",
    "details.errorMessage": "Error: {message}"
};

export const MESSAGES_BY_LOCALE: Record<"en" | "de", Record<string, string>> = {
    "en": MESSAGES_EN,
    "de": MESSAGES_DE
};

export function ErrorScreen(props: { intl: PackageIntl; error: Error }) {
    const intl = props.intl;
    return (
        <VStack padding={4} width="100%" height="100%" minWidth="350px" minHeight="150px">
            <Heading>{intl.formatMessage({ id: "title" })}</Heading>
            <Alert.Root status="error" size="lg" maxWidth={550} borderRadius="4px" mt={2}>
                <Alert.Indicator />
                <Alert.Content>
                    <Alert.Title fontWeight="bold">
                        {intl.formatMessage({ id: "alertTitle" })}
                    </Alert.Title>
                    <Alert.Description>
                        {intl.formatMessage({ id: "alertDescription" })}
                    </Alert.Description>
                </Alert.Content>
            </Alert.Root>

            {import.meta.env.DEV && (
                <Box
                    className="error-details"
                    mt={4}
                    p={2}
                    maxWidth="100%"
                    border={"dotted 2px"}
                    borderColor={"red.600"}
                    borderRadius={"4px"}
                >
                    <Text color={"red.600"} fontWeight={"bold"}>
                        {intl.formatMessage({ id: "details.title" }) + ":"}
                    </Text>
                    <ErrorRenderer {...props} />
                </Box>
            )}
        </VStack>
    );
}

/**
 * Renders a single error instance.
 */
function ErrorRenderer(props: { error: Error; intl: PackageIntl }) {
    const { error, intl } = props;
    const message = error?.message || "Unknown";
    const stackTrace = error?.stack;
    const stackTraceDisplay = stackTrace && (
        <>
            <Box
                p="2"
                mt={1}
                mb={3}
                maxHeight="500px"
                whiteSpace="pre"
                wordBreak="keep-all"
                overflow="auto"
                border="1px solid grey"
                borderRadius="4px"
                color="gray.700"
                fontSize="12px"
                fontFamily="monospace"
            >
                {stackTrace}
            </Box>
            {intl.formatMessage({ id: "details.inspect" })}
        </>
    );
    return (
        <Box p={2} className={"error-screen-stack-trace"}>
            <Text fontStyle={"italic"}>
                {intl.formatMessage({ id: "details.errorMessage" }, { message })}
            </Text>
            {stackTraceDisplay}
        </Box>
    );
}
