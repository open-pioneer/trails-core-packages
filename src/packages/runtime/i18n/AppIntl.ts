// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createLogger, Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { ApplicationMetadata } from "../metadata";
import { createPackageIntl, PackageIntl } from "./PackageIntl";
import { getBrowserLocales, I18nConfig } from "./pick";
const LOG = createLogger("runtime:i18n");

/**
 * Represents i18n info for the entire application.
 * Currently not exposed to user code.
 */
export interface AppIntl {
    /** Chosen locale */
    readonly locale: string;

    /** Supported locales from app metadata. */
    readonly supportedMessageLocales: string[];

    /** True if the locale can be used in this application (i.e. if there are any messages). */
    supportsLocale(locale: string): boolean;

    /** Given the package name, constructs a package i18n instance. */
    createPackageI18n(packageName: string): PackageIntl;
}

/**
 * Initializes the application's locale and fetches the appropriate i18n messages.
 */
export async function initI18n(
    appMetadata: ApplicationMetadata | undefined,
    forcedLocale: string | undefined
): Promise<AppIntl> {
    const messageLocales = appMetadata?.locales ?? [];
    const userLocales = getBrowserLocales();
    if (LOG.isDebug()) {
        LOG.debug(
            `Attempting to pick locale for user (locales: ${userLocales.join(
                ", "
            )}) from app (supported locales: ${messageLocales.join(
                ", "
            )}) [forcedLocale=${forcedLocale}].`
        );
    }

    const i18nConfig = new I18nConfig(messageLocales);
    const { locale, messageLocale } = i18nConfig.pickSupportedLocale(forcedLocale, userLocales);

    if (LOG.isDebug()) {
        LOG.debug(`Using locale '${locale}' with messages from locale '${messageLocale}'.`);
    }

    let messages: Record<string, Record<string, string>>;
    if (messageLocales.includes(messageLocale)) {
        try {
            messages = (await appMetadata?.loadMessages?.(messageLocale)) ?? {};
        } catch (e) {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to load messages for locale '${messageLocale}'.`,
                {
                    cause: e
                }
            );
        }
    }
    return {
        locale,
        supportedMessageLocales: messageLocales,
        supportsLocale(locale) {
            return i18nConfig.supportsLocale(locale);
        },
        createPackageI18n(packageName) {
            const packageMessage = messages?.[packageName] ?? {};
            return createPackageIntl(locale, packageMessage);
        }
    };
}
