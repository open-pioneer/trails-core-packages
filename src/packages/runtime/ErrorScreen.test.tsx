// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0

import { expect, it } from "vitest";
import { ErrorScreen, MESSAGES_BY_LOCALE } from "./ErrorScreen";
import { render, screen } from "@testing-library/react";
import { createPackageIntl } from "./i18n";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";

it("should successfully create an error screen component", async () => {
    const intl = createPackageIntl("en", MESSAGES_BY_LOCALE["en"]);
    const error = { name: "TestError", message: "test" };
    render(
        <PackageContextProvider>
            <div data-testid="error-screen">
                <ErrorScreen intl={intl} error={error} />
            </div>
        </PackageContextProvider>
    );
    const errorScreen = await screen.findByTestId("error-screen");

    expect(errorScreen).toMatchSnapshot();
});

it("should successfully create an error screen component with a stack trace", async () => {
    const intl = createPackageIntl("en", MESSAGES_BY_LOCALE["en"]);
    const error = { name: "TestError", message: "test", stack: "test stack trace" };
    render(
        <PackageContextProvider>
            <ErrorScreen intl={intl} error={error} />
        </PackageContextProvider>
    );
    const stackTraceDiv = document.getElementsByClassName("error-screen-stack-trace")[0];

    expect(stackTraceDiv).toBeDefined();
    expect(stackTraceDiv).toMatchSnapshot();
});
