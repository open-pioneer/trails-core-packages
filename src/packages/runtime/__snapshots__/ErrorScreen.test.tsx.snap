// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create an error screen component 1`] = `
<div
  data-testid="error-screen"
>
  <div
    class="css-er2cov"
    data-theme="light"
  >
    <div
      class="chakra-stack css-1thqo0p"
      data-theme="light"
    >
      <h2
        class="chakra-heading css-18j379d"
        data-theme="light"
      >
        Application start failed
      </h2>
      <div
        class="chakra-alert css-1v22m0m"
        data-status="error"
        data-theme="light"
        role="alert"
      >
        <span
          class="chakra-alert__icon css-14ogjxt"
          data-status="error"
          data-theme="light"
        >
          <svg
            class="chakra-icon css-qk6lof"
            data-theme="light"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
              fill="currentColor"
            />
          </svg>
        </span>
        <div
          class="chakra-alert__title css-tidvy5"
          data-status="error"
          data-theme="light"
        >
          Error
        </div>
        <div
          class="chakra-alert__desc css-zzks76"
          data-status="error"
          data-theme="light"
        >
          Unfortunately an error occurred during application start.
        </div>
      </div>
      <div
        class="error-details css-12gngag"
        data-theme="light"
      >
        <p
          class="chakra-text css-1pf3nin"
          data-theme="light"
        >
          Error details:
        </p>
        <div
          class="error-screen-stack-trace css-1syq16t"
          data-theme="light"
        >
          <p
            class="chakra-text css-1msjh1x"
            data-theme="light"
          >
            Error: test
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should successfully create an error screen component with a stack trace 1`] = `
<div
  class="error-screen-stack-trace css-1syq16t"
  data-theme="light"
>
  <p
    class="chakra-text css-1msjh1x"
    data-theme="light"
  >
    Error: test
  </p>
  <div
    class="css-1xl4qo1"
    data-theme="light"
  >
    test stack trace
  </div>
  Inspect the console to see more details.
</div>
`;
